File: address-poisoning-py/src/findings.py
from forta_agent import Finding, FindingSeverity, FindingType, EntityType

    
class AddressPoisoningFinding:

    def create_finding(transaction_event, anomaly_score, log_length, attackers, victims, alert_type):

        alert_description = {
            "ADDRESS-POISONING-ZERO-VALUE": "zero value",
            "ADDRESS-POISONING-LOW-VALUE": "low value",
            "ADDRESS-POISONING-FAKE-TOKEN": "fake token"
        }

        finding = Finding(
                    {
                        "name": "Possible Address Poisoning",
                        "description": f"Possible {alert_description[alert_type]} address poisoning transaction triggered by eoa - {transaction_event.from_} calling contract - {transaction_event.to}",
                        "alert_id": alert_type,
                        "type": FindingType.Suspicious,
                        "severity": FindingSeverity.Medium,
                        "metadata": {
                            "phishing_eoa": transaction_event.from_,
                            "phishing_contract": transaction_event.to,
                            "logs_length": log_length,
                            "attacker_addresses": attackers,
                            "victim_addresses": victims,
                            "anomaly_score": anomaly_score
                        },
                        "labels": [
                            {
                                "entityType": EntityType.Address,
                                "entity": transaction_event.from_,
                                "label": "attacker-eoa",
                                "confidence": 0.7
                            },
                            {
                                "entityType": EntityType.Address,
                                "entity": transaction_event.to,
                                "label": "attacker-contract",
                                "confidence": 0.7
                            },
                            {
                                "entityType": EntityType.Transaction,
                                "entity": transaction_event.hash,
                                "label": "address-poisoning",
                                "confidence": 0.7
                            },
                        ]
                    }  
                )

        return finding

File: suspicious-minting-ts/src/findings.ts
import {
  EntityType,
  Finding,
  FindingSeverity,
  FindingType,
  Label,
} from "forta-agent";

export const createFinding = (
  token: string,
  usdValue: string,
  txHash: string,
  mintRecipient: string,
  severity: FindingSeverity,
  txFrom: string
): Finding => {
  let labels: Label[] = [];
  let metadata: {
    [key: string]: string;
  } = {};
  metadata["initiator"] = txFrom;
  metadata["token"] = token;
  metadata["usdValue"] = usdValue;
  metadata["txHash"] = txHash;
  metadata["mintRecipient"] = mintRecipient;

  labels.push(
    Label.fromObject({
      entity: mintRecipient,
      entityType: EntityType.Address,
      label: "Attacker",
      confidence:
        severity === FindingSeverity.High
          ? 0.7
          : severity === FindingSeverity.Medium
          ? 0.6
          : 0.5,
      remove: false,
    })
  );

  labels.push(
    Label.fromObject({
      entity: txHash,
      entityType: EntityType.Transaction,
      label: "Attack",
      confidence:
        severity === FindingSeverity.High
          ? 0.7
          : severity === FindingSeverity.Medium
          ? 0.6
          : 0.5,
      remove: false,
    })
  );

  return Finding.fromObject({
    name: "Suspicious Mint",
    description:
      severity === FindingSeverity.High
        ? `Token mint of >$50k to ${mintRecipient} detected`
        : severity === FindingSeverity.Medium
        ? `Token mint of >$10k to new EOA ${mintRecipient} detected`
        : `Token mint of unknown value to new EOA ${mintRecipient} detected`,
    alertId:
      severity === FindingSeverity.High
        ? "SUSPICIOUS-MINT-1"
        : severity === FindingSeverity.Medium
        ? "SUSPICIOUS-MINT-2"
        : "SUSPICIOUS-MINT-3",
    severity,
    type: FindingType.Suspicious,
    metadata,
    labels,
  });
};
